File: 2016/day16.py
# @formatter:off
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day16.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file()
# @formatter:on

from utils import parsefile

init_state = parsefile(file_name, "")


def step(state):
	# copy state
	new_state = state[:]

	# reverse chars
	new_state = new_state[::-1]

	# swap all 1 and 0's around
	for i in range(len(new_state)):
		if new_state[i] == "0":
			new_state[i] = "1"
		else:
			new_state[i] = "0"

	return state + ["0"] + new_state


def checksum(state):
	check = [""] * (len(state) // 2)

	for i in range(0, len(state), 2):
		if state[i] == state[i + 1]:
			check[i // 2] = "1"
		else:
			check[i // 2] = "0"

	# print(check)

	if len(check) % 2 == 0:
		return checksum(check)

	return check


def randomise(size):
	state = init_state[:]
	while len(state) < size:
		state = step(state)

	check = checksum(state[:size])

	return "".join(check)


def part1():
	return randomise(272)


def part2():
	return randomise(35651584)


p1()
p2()

File: cpp/2016/day16.cpp
#include "../aocHelper.h"

class Day16 : public BaseDay
{
public:
	Day16() : BaseDay("16") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		// parse input
		string input_string;

		while (*input != '\n')
		{
			input_string += *input;
			input++;
		}

		auto step = [](string& state)
		{
			state += '0';

			// swap zeros and ones, but in reverse
			for (int i = state.length() - 2; i >= 0; i--)
			{
				if (state[i] == '0')
				{
					state += '1';
				}
				else
				{
					state += '0';
				}
			}
		};

		function<string(string::iterator, string::iterator, long long)> checksum;
		checksum = [&checksum](string::iterator start, string::iterator end, long long size) -> string
		{
			string cs;

			for (int i = 0; i < size; i += 2)
			{
				if (*(start + i) == *(start + i + 1))
				{
					cs += '1';
				}
				else
				{
					cs += '0';
				}
			}

			if (cs.length() % 2 == 0)
			{
				return checksum(cs.begin(), cs.end(), cs.length());
			}

			return cs;
		};

		auto randomise = [&step, &checksum](string& init_state, long long size) -> string
		{
			string state{ init_state };
			while (state.length() < size)
			{
				step(state);
			}

			string check = checksum(state.begin(), state.begin() + size, size);

			return check;
		};

		// part 1
		string p1 = randomise(input_string, 272);
		memcpy(this->stringResult.first, p1.c_str(), p1.length());

		// part 2
		string p2 = randomise(input_string, 35651584);
		memcpy(this->stringResult.second, p2.c_str(), p2.length());

		return { part1, part2 };
	}
};
